src/metabase/lib/portable.cljc

```diff
@@ -0,0 +1,124 @@
+(ns metabase.lib.portable
+  (:require
+   [medley.core :as m]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.lib.walk :as lib.walk]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]))
+
+(defprotocol ExportVisitor
+  (export-database     [this mp context database-id])
+  (export-source-table [this mp context source-table-id])
+  (export-source-card  [this mp context source-card-id])
+  (export-field        [this mp context field-id]))
+
+(defn export-visitor?
+  "Does x implement the ExportVisitor protocol?"
+  [x]
+  #?(:clj  (extends? ExportVisitor (class x))
+     :cljs (satisfies? ExportVisitor x)))
+
+(mr/def ::export-visitor
+  "Schema for things that satisfy the [[metabase.lib.portable/ExportVisitor]] protocol."
+  [:fn
+   {:error/message "Valid ExportVisitor"}
+   #'export-visitor?])
+
+(defprotocol ImportVisitor
+  (import-database     [this mp context database-info])
+  (import-source-table [this mp context source-table-info])
+  (import-source-card  [this mp context source-card-info])
+  (import-field        [this mp context field-info]))
```

Comment:
This, and all the functions following, are perfectly isomorphic of each other AFAICT. I think you can cut this namespace in half, by merging them into

```clojure
(defprotocol QueryVisitor
  (visit-database-id [this mp context database-id-form])
  (visit-table-id    [this mp context table-id-form])
  ...)
```

and all the functions likewise.